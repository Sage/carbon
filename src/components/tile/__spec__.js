import React from 'react';
import TestUtils from 'react/lib/ReactTestUtils';
import Tile from './tile';

describe('Tile', () => {
  let instance;

  describe('render', () => {
    it('renders div with a class of ui-tile', () => {
      instance = TestUtils.renderIntoDocument(<Tile>foo</Tile>);
      let podNode = TestUtils.scryRenderedDOMComponentsWithTag(instance, 'div')[0]
      expect(podNode.className).toEqual('ui-tile');
    });

    it('renders div with custom class', () => {
      instance = TestUtils.renderIntoDocument(<Tile className="foo">foo</Tile>);
      let podNode = TestUtils.scryRenderedDOMComponentsWithTag(instance, 'div')[0]
      expect(podNode.className).toEqual('ui-tile foo');
    });

    it('renders its children', () => {
      instance = TestUtils.renderIntoDocument(<Tile>foo</Tile>);
      let podNode = TestUtils.scryRenderedDOMComponentsWithTag(instance, 'div')[0]
      expect(podNode.textContent).toEqual('foo');
    });
  });
});
