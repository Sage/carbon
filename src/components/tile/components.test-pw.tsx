import React from "react";
import {
  ResponsiveCell,
  ResponsiveTileContainer,
  Tile,
  TileContent,
  TileFooter,
  TileFooterProps,
  TileProps,
} from ".";
import { Dl, Dt, Dd, DlProps } from "../definition-list";
import Accordion from "../accordion/accordion.component";
import Box from "../box/box.component";
import Button from "../button/button.component";
import Typography from "../typography/typography.component";

export const TileComponent = (props: TileProps) => {
  return (
    <Box height="180px">
      <Tile {...props}>
        <TileContent>Test Body One</TileContent>
        <TileContent>Test Body Two</TileContent>
        <TileContent>Test Body Three</TileContent>
      </Tile>
    </Box>
  );
};

export const TileFooterComponent = (props: Partial<TileFooterProps>) => {
  return (
    <Tile p={0} orientation="vertical">
      <Accordion p={0} borders="none" title="Accordion">
        <Dl dtTextAlign="left" ddTextAlign="right">
          <Dt>Coffee Subscription</Dt>
          <Dd>£7.00 a month</Dd>
          <Dt>Grind Size</Dt>
          <Dd>Espresso</Dd>
          <Dt>Quantity</Dt>
          <Dd>3kg</Dd>
          <Dd>
            <Button
              buttonType="tertiary"
              href="https://goo.gl/maps/GMReLoBpbn9mdZVZ7"
            >
              Have a promo code?
            </Button>
          </Dd>
        </Dl>
      </Accordion>
      <TileFooter {...props} p={3}>
        {props.children || (
          <>
            <Typography pr={2} display="inline" variant="b">
              Example footer text
            </Typography>
            <Typography display="inline">Example text</Typography>
          </>
        )}
      </TileFooter>
    </Tile>
  );
};

export const DlTileComponent = (props: Partial<DlProps>) => {
  return (
    <Tile>
      <Dl data-element="dl" {...props}>
        <Dt>Coffee Subscription</Dt>
        <Dd data-element="dd">£7.00 a month</Dd>
        <Dt>Grind Size</Dt>
        <Dd>Espresso</Dd>
        <Dt>Quantity</Dt>
        <Dd>3kg</Dd>
      </Dl>
    </Tile>
  );
};

export const TileComponentWithFalsyChildren = () => {
  return (
    <Tile>
      <TileContent data-element="tile-content">Foo</TileContent>
      <TileContent data-element="tile-content">{null}</TileContent>
      {false && <TileContent data-element="tile-content">Bar</TileContent>}
    </Tile>
  );
};

export const ResponsiveTile = () => {
  return (
    <Tile m={0} py={0}>
      <ResponsiveTileContainer>
        <ResponsiveCell py={2}>Test Body One</ResponsiveCell>
        <ResponsiveCell py={2}>Test Body Two</ResponsiveCell>
        <ResponsiveCell py={2}>
          Test Body Three With a very very long text
        </ResponsiveCell>
      </ResponsiveTileContainer>
    </Tile>
  );
};

export const ResponsiveTileContainerWithFalsyChildren = () => {
  return (
    <>
      <ResponsiveTileContainer />
      <ResponsiveTileContainer>Foo</ResponsiveTileContainer>
      <ResponsiveTileContainer>{null}</ResponsiveTileContainer>
      {false && <ResponsiveTileContainer>Bar</ResponsiveTileContainer>}
    </>
  );
};

export const ResponsiveCellWithFalsyChildren = () => {
  return (
    <>
      <ResponsiveCell />
      <ResponsiveCell>Foo</ResponsiveCell>
      <ResponsiveCell>{null}</ResponsiveCell>
      {false && <ResponsiveCell>Bar</ResponsiveCell>}
    </>
  );
};
