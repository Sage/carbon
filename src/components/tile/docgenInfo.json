{
  "src/__experimental__/components/tile/tile.component.js": [
    {
      "description": "",
      "displayName": "Tile",
      "methods": [],
      "props": {
        "as": {
          "type": {
            "name": "enum",
            "computed": true,
            "value": "OptionsHelper.tileThemes"
          },
          "required": false,
          "description": "Sets the theme of the tile - either 'tile' or 'transparent'",
          "defaultValue": {
            "value": "'tile'",
            "computed": false
          }
        },
        "children": {
          "type": {
            "name": "node"
          },
          "required": false,
          "description": "The content to render within the tile. Each child will be wrapped with\na TileContent wrapper, which allows any individual child component to take a\npercentage-based width prop, dictating the percentage of the tile width it will take up.\n\nWidth will have no effect on a child component if the tile orientation is set to 'vertical'."
        },
        "orientation": {
          "type": {
            "name": "enum",
            "computed": true,
            "value": "OptionsHelper.orientation"
          },
          "required": false,
          "description": "The orientation of the tile - set to either horizontal or vertical",
          "defaultValue": {
            "value": "'horizontal'",
            "computed": false
          }
        },
        "padding": {
          "type": {
            "name": "enum",
            "computed": true,
            "value": "OptionsHelper.sizesPod"
          },
          "required": false,
          "description": "Sets the Tile padding. Accepted values are:\n'extra-small' = 8px\n'small' = 16px\n'medium' = 24px\n'large' = 32px\n'extra-large' = 40px",
          "defaultValue": {
            "value": "'medium'",
            "computed": false
          }
        },
        "width": {
          "type": {
            "name": "number"
          },
          "required": false,
          "description": "Set a percentage-based width for the whole Tile component, relative to its parent.\nIf unset or zero, this will default to 100%."
        }
      }
    }
  ]
}
