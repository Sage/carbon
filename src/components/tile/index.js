export { default } from './tile.component';
