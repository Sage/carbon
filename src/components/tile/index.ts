export { default as Tile } from "./tile.component";
export type { TileProps } from "./tile.component";
export { default as TileContent } from "./tile-content";
export type { TileContentProps } from "./tile-content";
export { default as TileFooter } from "./tile-footer";
export type { TileFooterProps } from "./tile-footer";
export { default as ResponsiveTileContainer } from "./flow-tile-container";
export type { FlowTileContainerProps } from "./flow-tile-container";
export { default as ResponsiveCell } from "./flow-cell";
export type { FlowCellProps } from "./flow-cell";
