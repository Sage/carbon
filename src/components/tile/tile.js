import React from 'react';
import classNames from 'classnames';

/**
 * @class Tile
 * @constructor
 */
class Tile extends React.Component {
  static propTypes = {
    /**
     * A required prop. This is what the Tile will display.
     *
     * @property children
     * @type {Multiple}
     */
    children: React.PropTypes.node.isRequired
  }

  /**
   * Main class getter
   *
   * @method mainClasses
   * @return {void}
   */
  get mainClasses() {
    return classNames("ui-tile", this.props.className);
  }

  /**
   * @method render
   * @return {Object} JSX
   */
  render() {
    return (
      <div className={ this.mainClasses }>
        { this.props.children }
      </div>
    );
  }
}

export default Tile;
