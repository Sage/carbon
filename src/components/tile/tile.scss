.ui-tile {
  background-color: #FFF;
  border-radius: 4px;
  box-shadow: 0 1px 3px 0 rgba(0, 0, 0, 0.2);
}
