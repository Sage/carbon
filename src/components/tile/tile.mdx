import { Meta, ArgTypes, Canvas } from "@storybook/blocks";

import * as DlStories from "../definition-list/definition-list.stories";
import * as DtStories from "../definition-list/dt/dt.stories";
import * as DdStories from "../definition-list/dd/dd.stories";
import * as FlexTileContainerStories from "./flex-tile-container/flex-tile-container.stories";
import * as FlexTileCellStories from "./flex-tile-cell/flex-tile-cell.stories";
import * as TileFooterStories from "./tile-footer/tile-footer.stories";
import * as TileHeaderStories from "./tile-header/tile-header.stories";
import * as TileStories from "./tile.stories";

<Meta title="Tile" of={TileStories} />

# Tile

Tiles are containers for static content such as text, and don't contain controls.

## Contents

- [Quick Start](#quick-start)
- [Examples](#examples)
- [Props](#props)

## Quick Start

```javascript
import {
  Tile,
  TileContent,
  TileFooter,
  FlexTileContainer,
  FlexTileCell,
} from "carbon-react/lib/components/tile";
```

## Designer notes

- Tiles should be reserved for static content. If you require a container with controls or interactive content, use the [Card component](../?path=/docs/card--docs) instead.

## Examples

### Default

By default, the Tile component will have a horizontal layout.

<Canvas of={TileStories.DefaultStory} />

### Roundness variants for corners

<Canvas of={TileStories.LargeRoundness} />

<Canvas of={TileStories.SmallRoundness} />

### Vertical

The Tile component can be also have a vertical layout.

<Canvas of={TileStories.Vertical} />

### With TileFooter

<Canvas of={TileStories.WithTileFooter} />

### With TileHeader

<Canvas of={TileStories.WithTileHeader} />

### With Button in TileHeader

<Canvas of={TileStories.WithButtonInTileHeader} />

### With custom widths

The Tile component can also have a custom `width` which accepts either numbers or strings: any decimal between 0 and 1.0
will be converted into a percentage, any number greater will be be treated as pixels and any valid CSS string will also work.
Any child wrapped by a Tile component can be passed an optional `width` prop if its interface supports it.

<Canvas of={TileStories.CustomWidths} />

### With custom heights

The Tile component can also have a custom `height` which accepts either numbers or strings: any decimal between 0 and 1.0
will be converted into a percentage, any number greater will be be treated as pixels and any valid CSS string will also work.
Any child wrapped by a Tile component can be passed an optional `height` prop if its interface supports it.

<Canvas of={TileStories.CustomHeights} />

### Active Tile

<Canvas of={TileStories.Active} />

### Grey variant

<Canvas of={TileStories.Grey} />

### Custom borders

You can set the `borderVariant` property to alter the border colour as well as the border width with the `borderWidth` property.

<Canvas of={TileStories.CustomBorders} />

### With inline styling

The Tile component can also have inline styling applied to the content.

<Canvas of={TileStories.WithInline} />

### With different padding and margin

<Canvas of={TileStories.WithDifferentPaddingAndMargin} />

### With Definition List

The Definition List component should only be used inside of the Tile component. The purpose of the Definition List component is to mimic
the same tags used in HTML5 but instead they are React elements.

Please note that Dl, Dd and Dt use our common spacing props.

<Canvas of={TileStories.WithDefinitionListDefault} />

### With Definition List and custom `StyledDtDiv` width

Providing a number value to the `w` prop, sets the width of the `<Dt />` element as a percentage. The remaining percentage is then assigned to `<Dd />` to occupy the available space.

<Canvas of={TileStories.WithDefinitionListAndCustomWidth} />

### With Definition List with custom text alignment

In this example `dtTextAlign` has been provided the string 'left' and `ddTextAlign` has been provided the string 'right'.

<Canvas of={TileStories.WithDefinitionListAndCustomTextAlignment} />

### With Definition List and ActionPopover and Icon support

<Canvas of={TileStories.WithDefinitionListAndActionPopoverAndIconSupport} />

### With Accordion and Definition List

<Canvas of={TileStories.WithAccordion} />

### With Accordion and TileFooter

<Canvas of={TileStories.WithAccordionAndTileFooter} />

### Responsive Tile

To make the content of a tile responsive, use `FlexTileCell` components inside a `FlexTileContainer`.
The content of a `FlexTileCell` can be what you want.

When the size of the screen is too small to display all the cells on one line, some will be automatically wrapped to a new line.
If you add a `FlexTileDivider` in each cell, a divider will appear between the lines.

<Canvas of={TileStories.ResponsiveDefaultStory} />

### Responsive Tile with custom gaps

<Canvas of={TileStories.CustomGaps} />

### Responsive Tile with fixed width for cells

With `flexGrow={0}`, the width of each cell will be defined by `flexBasis`. It will not grow or shrink according to the remaining space.

<Canvas of={TileStories.FixedContainers} />

### Responsive Tile with flex width for cells

If `flexGrow` is **not** `0`, the width (= `flexBasis`) of each cell will be a min-width. It will grow to fill the remaining space and shrink until reaching that min-width.
You can use `maxWidth` to constrain the growth.

<Canvas of={TileStories.FlexContainers} />

### Responsive Tile with proportionate widths

Responsive cells can be set so that they have a proportionate width to other cells in the same tile.

Use this values on `flexGrow`:

- 1: The remaining space in the container will be distributed equally to all cells.
- 2: That cell will try to take up twice as much of the space than the others.
- 3: That cell will try to take up thrice as much of the space than the others.
- ...

<Canvas of={TileStories.ProportionateWidths} />

### Responsive Tile with align content

Use `justifyContent` to place your content inside the width allocated to the cell.

<Canvas of={TileStories.Align} />

### Tile

<ArgTypes of={TileStories} />

### TileFooter

<ArgTypes of={TileFooterStories} />

### TileHeader

<ArgTypes of={TileHeaderStories} />

### Dl

<ArgTypes of={DlStories} />

### Dd

<ArgTypes of={DdStories} />

### Dt

<ArgTypes of={DtStories} />

### FlexTileContainer

<ArgTypes of={FlexTileContainerStories} />

### FlexTileCell

<ArgTypes of={FlexTileCellStories} />
