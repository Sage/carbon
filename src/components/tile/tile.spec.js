import React from 'react';
import TestRenderer from 'react-test-renderer';
import 'jest-styled-components';
import { mount } from 'enzyme';
import { css } from 'styled-components';
import Tile from '.';
import { TileContent } from './tile.style';
import Content from '../content';
import { assertStyleMatch } from '../../__spec_helper__/test-utils';

function render(props, renderer = TestRenderer.create) {
  return renderer(
    <Tile { ...props }>
      <Content key='one'>Child 1</Content>
      <Content>Child 2</Content>
    </Tile>
  );
}

describe('Tile', () => {
  it('renders base styles', () => {
    const wrapper = render({});

    expect(wrapper).toMatchSnapshot();
  });

  describe('wrapping of children in TileContent components', () => {
    describe('standard', () => {
      const wrapper = render({}, mount);
      const tileContents = wrapper.find(TileContent).getElements();

      it('contains one TileContent for each child', () => {
        expect(tileContents.length).toBe(2);
      });

      it.each([0, 1])('TileContent[%i] contains the passed Content as its own child', (childIndex) => {
        expect(tileContents[childIndex].props.children.type.name).toBe('Content');
        expect(tileContents[childIndex].props.children.props.children).toBe(`Child ${childIndex + 1}`);
      });
    });

    describe('when something causes a child element to return nothing', () => {
      const children = [
        <Content key='one'>Child 1</Content>,
        undefined
      ];

      const wrapper = mount(
        <Tile>
          {children}
        </Tile>
      );

      const tileContents = wrapper.find(TileContent).getElements();

      it('only contains one TileContent', () => {
        expect(tileContents.length).toBe(1);
      });
    });
  });

  describe('styles', () => {
    describe('as', () => {
      it('renders a white background when as === "tile"', () => {
        const wrapper = render({ as: 'tile' }).toJSON();

        assertStyleMatch({ backgroundColor: '#FFFFFF' }, wrapper);
      });

      it('renders a transparent background when as === "transparent"', () => {
        const wrapper = render({ as: 'transparent' }).toJSON();

        assertStyleMatch({ backgroundColor: 'transparent' }, wrapper);
      });
    });

    describe('orientation', () => {
      describe('when it is horizontal', () => {
        const wrapper = render({ orientation: 'horizontal' }).toJSON();

        it('sets the correct flex-direction on the main wrapper', () => {
          assertStyleMatch({ flexDirection: 'row' }, wrapper);
        });

        it('sets the TileContent display to inline', () => {
          assertStyleMatch({ display: 'inline' }, wrapper, { modifier: css`${TileContent}` });
        });

        it('sets padding-right for all but the last TileContent', () => {
          assertStyleMatch(
            {
              paddingRight: '24px'
            },
            wrapper,
            { modifier: css`${`${TileContent}:not(:last-of-type)`}` }
          );
        });

        it('sets border-left and padding-left for all but the first TileComponent', () => {
          assertStyleMatch(
            {
              borderLeft: 'solid 1px #E5EAEC',
              paddingLeft: '24px'
            },
            wrapper, { modifier: css`${`${TileContent} + ${TileContent}`}` }
          );
        });
      });

      describe('when it is vertical', () => {
        const wrapper = render({ orientation: 'vertical' }).toJSON();

        it('sets the correct flex-direction on the main wrapper', () => {
          assertStyleMatch({ flexDirection: 'column' }, wrapper);
        });

        it('sets the TileContent width to auto', () => {
          assertStyleMatch({ width: 'auto' }, wrapper, { modifier: css`${TileContent}` });
        });

        it('sets padding-bottom for all but the last TileContent', () => {
          assertStyleMatch(
            { paddingBottom: '24px' },
            wrapper,
            { modifier: css`${`${TileContent}:not(:last-of-type)`}` }
          );
        });

        it('sets border-top and padding-top, and width: auto for all but the first TileComponent', () => {
          assertStyleMatch(
            {
              borderTop: 'solid 1px #E5EAEC',
              paddingTop: '24px'
            },
            wrapper, { modifier: css`${`${TileContent} + ${TileContent}`}` }
          );
        });
      });

      describe('width', () => {
        it('sets width to 100% when width prop is undefined', () => {
          const wrapper = render().toJSON();

          assertStyleMatch({ width: '100%' }, wrapper);
        });

        it('sets width to 100% when width prop is 0', () => {
          const wrapper = render({ width: 0 }).toJSON();

          assertStyleMatch({ width: '100%' }, wrapper);
        });

        it('sets width to the passed percentage value when width prop is non-zero', () => {
          const wrapper = render({ width: 25 }).toJSON();

          assertStyleMatch({ width: '25%' }, wrapper);
        });

        it('is overridden when the pixelWidth prop is set', () => {
          const wrapper = render({ pixelWidth: 500, width: 25 }).toJSON();

          assertStyleMatch({ width: '500px' }, wrapper);
        });
      });

      describe('padding', () => {
        const paddingSizes = [
          ['XS', '8px'],
          ['S', '16px'],
          ['M', '24px'],
          ['L', '32px'],
          ['XL', '40px']
        ];

        it.each(paddingSizes)('when %s, padding is set to %s', (option, value) => {
          const wrapper = render({ padding: option }).toJSON();

          assertStyleMatch({ padding: value }, wrapper);
        });
      });
    });
  });

  describe('TileContent', () => {
    describe('styles', () => {
      function renderTileContent(props) {
        return TestRenderer.create(<TileContent { ...props }>Test</TileContent>).toJSON();
      }

      it('has the correct base styles', () => {
        const wrapper = renderTileContent();

        assertStyleMatch(
          {
            position: 'relative',
            flexGrow: '1',
            width: undefined
          },
          wrapper
        );
      });

      describe('width', () => {
        it('sets width to the passed percentage value and flex-grow to 0 when width prop is non-zero', () => {
          const wrapper = renderTileContent({ width: 25 });

          assertStyleMatch({ flexGrow: '0', width: '25%' }, wrapper);
        });
      });
    });
  });
});
