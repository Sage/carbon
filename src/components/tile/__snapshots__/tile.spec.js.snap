// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tile renders base styles 1`] = `
.c2 {
  position: relative;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c0 {
  background-color: #FFFFFF;
  border: 1px solid #CCD6DA;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  padding: 24px;
  position: relative;
  width: 100%;
}

.c0 .c1 {
  box-sizing: border-box;
  display: inline;
}

.c0 .c1:not(:last-of-type) {
  padding-right: 24px;
}

.c0 .c1 + .c1 {
  margin-top: 0;
  border-left: solid 1px #E5EAEC;
  padding-left: 24px;
}

.c3 .c1 {
  box-sizing: border-box;
  width: auto;
}

.c3 .c1:not(:last-of-type) {
  padding-bottom: 24px;
}

.c3 .c1 + .c1 {
  margin-top: 0;
  border-top: solid 1px #E5EAEC;
  padding-top: 24px;
}

<div
  className="c0"
  data-component="tile"
  orientation="horizontal"
>
  <div
    className="c1 c2"
  >
    <div
      className="carbon-content carbon-content--primary carbon-content--align-left"
      data-component="content"
    >
      <div
        className="carbon-content__title"
        data-element="title"
        style={Object {}}
      />
      <div
        className="carbon-content__body"
        data-element="body"
        style={Object {}}
      >
        Child 1
      </div>
    </div>
  </div>
  <div
    className="c1 c2"
  >
    <div
      className="carbon-content carbon-content--primary carbon-content--align-left"
      data-component="content"
    >
      <div
        className="carbon-content__title"
        data-element="title"
        style={Object {}}
      />
      <div
        className="carbon-content__body"
        data-element="body"
        style={Object {}}
      >
        Child 2
      </div>
    </div>
  </div>
</div>
`;
